### PR TITLE
[Spark runner] Removal of Spark 2 runner support

### LINKED ISSUE #25259: [Task]: Planned removal of Spark 2 runner support
### What needs to happen?

The runner for Spark 2 was deprecated quite a while back in August 2022 with the release of [Beam 2.41.0](https://github.com/apache/beam/blob/master/CHANGES.md#2410---2022-08-23). We’re planning to move ahead with this and finally remove support for Spark 2 (`beam-runners-spark`) to only maintain support for Spark 3 (`beam-runners-spark-3`) going forward.
 
Note, Spark 2.4.8 is the latest release of Spark 2. It was released in May 2021 and [no further releases](https://spark.apache.org/versioning-policy.html) of 2.4.x should be expected, even for bug fixes.
 

### Issue Priority

Priority: 2 (default / most normal work should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [X] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by mosche**
The runner for Spark 2 was deprecated quite a while back in August 2022 with the release of [Beam 2.41.0](https://github.com/apache/beam/blob/master/CHANGES.md#2410---2022-08-23). 

This PR finally removes support for Spark 2 (beam-runners-spark) to only maintain support for Spark 3 (beam-runners-spark-3) going forward.

Closes #25259 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by mosche** [comment]
R: @aromanenko-dev

**Comment by mosche** [comment]
R: @je-ik 
R: @JozoVilcek

**Comment by mosche** [comment]
Run Python_Runners PreCommit

**Comment by mosche** [comment]
Run Python_Runners PreCommit

**Comment by mosche** [comment]
Run Java_Pulsar_IO_Direct PreCommit

**Review by JozoVilcek — APPROVED** [review]
Hey @mosche , LGTM

**Comment by mosche** [comment]
@aromanenko-dev Could you have a look, pls. Any objections moving ahead with the removal?

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -730,8 +729,6 @@ class BeamModulePlugin implements Plugin<Project> {
         slf4j_jcl                                   : "org.slf4j:slf4j-jcl:$slf4j_version",
         snappy_java                                 : "org.xerial.snappy:snappy-java:1.1.8.4",
         spark_core                                  : "org.apache.spark:spark-core_2.11:$spark2_version",
```
Comment (aromanenko-dev) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:731]: Should this line and a one with `spark_streaming` below be removed as well?

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 2]
```diff
@@ -1535,9 +1535,8 @@ def _add_argparse_args(cls, parser):
     parser.add_argument(
         '--spark_version',
         default='3',
-        choices=['3', '2'],
-        help='Spark major version to use. '
-        'Note, Spark 2 support is deprecated')
+        choices=['3'],
```
Comment (aromanenko-dev) [inline sdks/python/apache_beam/options/pipeline_options.py:1538]: Do you keep it for future Spark versions, like `Spark 4`?

File: sdks/python/apache_beam/runners/portability/spark_runner.py [hunk 3]
```diff
@@ -93,9 +93,7 @@ def path_to_jar(self):
       return self._jar
     else:
       if self._spark_version == '2':
```
Comment (aromanenko-dev) [inline sdks/python/apache_beam/runners/portability/spark_runner.py:95]: Why we still need this check?

**Review by aromanenko-dev — APPROVED** [review]
Thanks, LGTM, just several minor questions, ptal

Before merging, please, run all related `pre-` and `post-` commit jobs to make sure that nothing is broken.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 1 above)
Comment (mosche) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:731]: These are still used for cdapio (spark receiver)

File: sdks/python/apache_beam/options/pipeline_options.py (hunk 2 above)
Comment (mosche) [reply sdks/python/apache_beam/options/pipeline_options.py:1538]: No, removing the pipeline option would be a breaking change for the ones using spark 3 and setting this explicitly.

File: sdks/python/apache_beam/runners/portability/spark_runner.py (hunk 3 above)
Comment (mosche) [reply sdks/python/apache_beam/runners/portability/spark_runner.py:95]: yes, if somebody explicitly requires spark 2 it should fail fast

**Comment by mosche** [comment]
Run Python Spark ValidatesRunner

**Comment by mosche** [comment]
Run Java PostCommit

**Comment by mosche** [comment]
Run SQL PostCommit

**Comment by mosche** [comment]
> Before merging, please, run all related pre- and post- commit jobs to make sure that nothing is broken.

@aromanenko-dev  This pretty much triggered all `pre` jobs automatically in addition to the `post` jobs run above. Anything else you have in mind?

**Comment by aromanenko-dev** [comment]
@mosche Nope, I think it should be enough

**Comment by aromanenko-dev** [comment]
@mosche Could you rebase a feature branch against `master` instead of merging?

**Comment by mosche** [comment]
@aromanenko-dev That's what github does if resolving a conflict in the UI. It absolutely doesn't make a difference if squashed before merging.

**Comment by mosche** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Pull request merged**